firedrake/output.py

```diff
@@ -651,3 +756,366 @@ def write(self, *functions, **kwargs):
+                                          PREFIX + "_mixed_function_spaces")
+            if name in self.h5pyfile[base_base_path]:
+                base_path = os.path.join(base_base_path, name)
+                n = self.get_attr(base_path, PREFIX + "_num_sub_spaces")
+                Vsub_list = []
+                for i in range(n):
+                    path = os.path.join(base_path, str(i))
+                    Vsub_mesh_name = self.get_attr(path, PREFIX + "_mesh")
+                    Vsub_name = self.get_attr(path, PREFIX + "_function_space")
+                    Vsub = self.load_function_space(Vsub_name,
+                                                    Vsub_mesh_name,
+                                                    distribution_parameters=distribution_parameters)
+                    Vsub_list.append(Vsub)
+                V = functools.reduce(lambda a, b: a * b, Vsub_list)
+                V.name = name
+                return V
+        dist_key = hash(frozenset(distribution_parameters.items()))
+        mesh_key = (mesh_name, dist_key)
+        fs_key = (name, ) + mesh_key
+        if fs_key in self._function_spaces:
+            return self._function_spaces[fs_key]
+        # Load mesh
+        mesh = self.load_mesh(mesh_name, distribution_parameters=distribution_parameters)
+        mesh.init()
+        # Load element
+        sectiondm_name = self.find_sectiondm_name(mesh_name, name)
+        path = os.path.join("/topologies",
+                            mesh.topology_dm.getName(),
+                            "dms",
+                            sectiondm_name,
+                            PREFIX + "_function_spaces",
+                            name)
+        element_b = self.get_attr(path, PREFIX + "_ufl_element").tobytes()
+        element = pickle.loads(element_b)
+        # Load shared data
+        tmesh = mesh.topology
+        dm = tmesh.topology_dm
+        sectiondm = PETSc.DMShell().create(comm=dm.comm)
+        sectiondm.setName(sectiondm_name)
+        sectiondm.setPointSF(dm.getPointSF())
+        section = PETSc.Section().create(comm=dm.comm)
+        section.setPermutation(tmesh._plex_renumbering)
+        sectiondm.setSection(section)
+        sfXB = self._meshes[mesh_key]["sfXB"]
+        sfXC = self._meshes[mesh_key].setdefault("sfXC",
+                                                 sfXB.compose(tmesh.sfBC) if tmesh.sfBC else sfXB)
+        gsf, lsf = dm.sectionLoad(self.viewer, sectiondm, sfXC)
+        sd_key = get_shared_data_key_for_checkpointing(tmesh, element)
+        _ = get_global_numbering(tmesh, sd_key, global_numbering=sectiondm.getSection())
+        # Construct function space
+        V = FunctionSpace(mesh, element, name=name)
+        self._function_load_utils[sd_key + mesh_key] = (sectiondm, gsf, lsf)
+        self._function_spaces[fs_key] = V
+        return V
+
+    def load_function(self, name, mesh_name=DEFAULT_MESH_NAME, distribution_parameters={}):
```

Comment:
Why do I need distribution parameters when loading a function?

Reply:
I mean, I know why, but why is this the interface?

Could we instead have a loading scheme that is:

```
with ... as f:
   mesh = f.load_mesh(...)
   g = f.load_function(name, mesh)
```

Reply:
Yes. Updated. Thanks. I think this is intuitive.